client/dashboard/components/settings/general/store-address.js

```diff
@@ -75,57 +83,75 @@ export function getCountryStateOptions() {
  * Get the autofill countryState fields and set value from filtered options.
  *
  * @param {Array} options Array of filterable options.
- * @param {String} countryState The value of the countryState field.
+ * @param {string} countryState The value of the countryState field.
  * @param {Function} setValue Set value of the countryState input.
+ * @param {Function} _useState React useState.
+ * @param {Function} _useEffect React useEffect.
  * @return {Object} React component.
  */
-export function getCountryStateAutofill( options, countryState, setValue ) {
-	const [ autofillCountry, setAutofillCountry ] = useState( '' );
-	const [ autofillState, setAutofillState ] = useState( '' );
-
-	useEffect(
-		() => {
-			let filteredOptions = [];
-			const countrySearch = new RegExp( escapeRegExp( autofillCountry ), 'i' );
-			if ( autofillState.length || autofillCountry.length ) {
-				filteredOptions = options.filter( option => countrySearch.test( option.label ) );
-			}
-			if ( autofillCountry.length && autofillState.length ) {
-				const stateSearch = new RegExp( escapeRegExp( autofillState.replace( /\s/g, '' ) ), 'i' );
-				filteredOptions = filteredOptions.filter( option =>
-					stateSearch.test( option.label.replace( '-', '' ).replace( /\s/g, '' ) )
+export function getCountryStateAutofill( options, countryState, setValue, _useState, _useEffect ) {
+	const [ autofillCountry, setAutofillCountry ] = _useState( '' );
+	const [ autofillState, setAutofillState ] = _useState( '' );
+
```

Comment:
Passing React Hooks as props is a bad idea because it potentially breaks the rule of hooks where order of execution matters. Because `getCountryStateAutofill` might be invoked in different order on renders, this will cause bugs.  Generally, _never_ pass hooks as props.

I'm suspecting you did this because of the following warning:

> React Hook "useState" is called in function "getCountryStateAutofill" which is neither a React function component or a custom React Hook function.eslint(react-hooks/rules-of-hooks)

The solution for this would be to convert `getCountryStateAutofill` to a functional component and/or custom hook.  I would simply convert to a custom react hook (which just requires the addition of the `use` keyword to the function):

```jsx
export function useCountryStateAutofill = ( options, countryState, setValue ) => {
  /** original code **/
}
```

Then you'll just have to refactor everywhere `getCountryStateAutofill` occurs to `useCountryStateAutofill` (just make sure you follow the rule of hooks in implementation). I think I only found one place it's used in code and it doesn't look like there should be issues with conversion of this method to a hook.

Reply:
Good call @nerrad , I wasn't aware of that.

https://github.com/woocommerce/woocommerce-admin/pull/3674/commits/ff890cfc9a87b122142444dd3c90b17101d39ba6 turns the function into a custom hook, which is only used in that file.